tests/framework/prototype.py

```diff
@@ -272,13 +316,109 @@ def wait(self):
                 namespace = self.namespace, label_selector='ray.io/node-type=worker')
             if (len(headpods.items) == 0 and len(workerpods.items) == 0):
                 converge = True
-                logger.info("--- RayClusterDeleteCREvent %s seconds ---", time.time() - start_time)
+                logger.info("--- Cleanup RayCluster %s seconds ---", time.time() - start_time)
+                break
+            time.sleep(1)
+        k8s_v1_api.api_client.rest_client.pool_manager.clear()
+        k8s_v1_api.api_client.close()
+        if not converge:
+            logger.info("RayClusterAddCREvent clean_up() failed to converge in %d seconds.",
+                self.timeout)
+            logger.info("expected_head_pods: 0, expected_worker_pods: 0")
+            show_cr_info(self.namespace)
+            raise Exception("RayClusterAddCREvent clean_up() timeout")
+
+class RayServiceAddCREvent(CREvent):
+    """CREvent for RayService addition"""
+    def exec(self):
+        """Wait for RayService to converge"""""
+        if not self.filepath:
+            k8s_cr_api = client.CustomObjectsApi()
+            k8s_cr_api.create_namespaced_custom_object(
+                group = 'ray.io',version = 'v1alpha1', namespace = self.namespace,
+                plural = 'rayservices', body = self.custom_resource_object)
+            k8s_cr_api.api_client.rest_client.pool_manager.clear()
+            k8s_cr_api.api_client.close()
+        else:
+            subprocess.run(f"kubectl apply -n {self.namespace} -f {self.filepath}",
+                shell = True, check = True)
+
+    def wait(self):
+        """Wait for RayService to converge"""""
+        start_time = time.time()
+        expected_head_pods = search_path(self.custom_resource_object,
+            "spec.rayClusterConfig.headGroupSpec.replicas".split('.'), default_value=1)
+        worker_group_specs = search_path(self.custom_resource_object,
+            "spec.rayClusterConfig.workerGroupSpecs".split('.'), default_value=[])
+        expected_worker_pods = 0
```

Comment:
Define a function to get `expected_worker_pods` for different CRD (RayCluster / RayJob / RayService)